File: mobile-rn/src/services/CDPAuthService.ts
import axios, { AxiosInstance, AxiosError } from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage'; // Common storage for RN

// --- Configuration ---
const API_BASE_URL = 'https://api.54link.io/cdp/v1';
const AUTH_TOKEN_KEY = '@CdpAuth:Token';
const REFRESH_TOKEN_KEY = '@CdpAuth:RefreshToken';

// --- Type Definitions for Request/Response Payloads ---

/**
 * Interface for the stored authentication tokens.
 */
interface AuthTokens {
  accessToken: string;
  refreshToken: string;
}

/**
 * Interface for a successful API response.
 * @template T The type of the data payload.
 */
interface ApiResponse<T> {
  success: true;
  data: T;
}

/**
 * Interface for an error API response.
 */
interface ApiErrorResponse {
  success: false;
  message: string;
  code?: string;
  details?: any;
}

/**
 * Type for all possible API responses.
 * @template T The type of the data payload for success.
 */
type ServiceResponse<T> = ApiResponse<T> | ApiErrorResponse;

// --- Authentication Payloads ---

interface SendOtpRequest {
  email: string;
}

interface VerifyOtpRequest {
  email: string;
  otp: string;
}

interface AuthSuccessResponse {
  accessToken: string;
  refreshToken: string;
  userId: string;
  walletCreated: boolean;
}

interface WalletCreationResponse {
  walletId: string;
  message: string;
}

// --- CDP Authentication Service Class ---

/**
 * A production-ready service class for handling all CDP authentication,
 * session management, and wallet creation logic in a React Native application.
 * It uses Axios for HTTP requests and AsyncStorage for secure token storage.
 */
export class CdpAuthService {
  private api: AxiosInstance;

  constructor() {
    // 1. Initialize Axios instance with base URL and interceptors
    this.api = axios.create({
      baseURL: API_BASE_URL,
      headers: {
        'Content-Type': 'application/json',
      },
      timeout: 15000, // 15 seconds timeout
    });

    // 2. Setup request interceptor to attach access token
    this.api.interceptors.request.use(
      async (config) => {
        const tokens = await this.getTokens();
        if (tokens?.accessToken) {
          config.headers.Authorization = `Bearer ${tokens.accessToken}`;
        }
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );

    // 3. Setup response interceptor for automatic token refresh
    this.api.interceptors.response.use(
      (response) => response,
      async (error: AxiosError) => {
        const originalRequest = error.config;
        // Check for 401 Unauthorized and ensure it's not a refresh token request loop
        if (error.response?.status === 401 && originalRequest && !(originalRequest as any)._retry) {
          (originalRequest as any)._retry = true; // Mark request as retried
          try {
            const newTokens = await this.refreshSession();
            if (newTokens) {
              // Update the Authorization header for the original request
              originalRequest.headers.Authorization = `Bearer ${newTokens.accessToken}`;
              // Re-run the original request with the new token
              return this.api(originalRequest);
            }
          } catch (refreshError) {
            // If refresh fails, clear session and force logout
            console.error('Token refresh failed, logging out:', refreshError);
            await this.clearSession();
            // Optionally, emit an event to notify the app to navigate to login screen
            // E.g., EventBus.emit('sessionExpired');
            return Promise.reject(refreshError);
          }
        }
        return Promise.reject(error);
      }
    );
  }

  // --- Utility Methods for Token Storage ---

  /**
   * Securely stores the access and refresh tokens.
   * @param tokens The tokens to store.
   */
  private async saveTokens(tokens: AuthTokens): Promise<void> {
    await AsyncStorage.setItem(AUTH_TOKEN_KEY, tokens.accessToken);
    await AsyncStorage.setItem(REFRESH_TOKEN_KEY, tokens.refreshToken);
  }

  /**
   * Retrieves the stored access and refresh tokens.
   * @returns A promise that resolves to the tokens or null if not found.
   */
  public async getTokens(): Promise<AuthTokens | null> {
    const accessToken = await AsyncStorage.getItem(AUTH_TOKEN_KEY);
    const refreshToken = await AsyncStorage.getItem(REFRESH_TOKEN_KEY);
    if (accessToken && refreshToken) {
      return { accessToken, refreshToken };
    }
    return null;
  }

  /**
   * Clears all stored session tokens.
   */
  public async clearSession(): Promise<void> {
    await AsyncStorage.removeItem(AUTH_TOKEN_KEY);
    await AsyncStorage.removeItem(REFRESH_TOKEN_KEY);
  }

  /**
   * Checks if a user is currently logged in (has valid tokens).
   * @returns A promise that resolves to true if logged in, false otherwise.
   */
  public async isLoggedIn(): Promise<boolean> {
    const tokens = await this.getTokens();
    // In a real app, you might also want to check token expiry here
    return !!tokens?.accessToken;
  }

  // --- Core Authentication Methods ---

  /**
   * Sends an OTP to the user's email for login or registration.
   * @param email The user's email address.
   * @returns A service response indicating success or failure.
   */
  public async sendOtp(email: string): Promise<ServiceResponse<{ message: string }>> {
    if (!email) {
      return { success: false, message: 'Email is required for OTP request.' };
    }
    try {
      const response = await this.api.post<ApiResponse<{ message: string }>>('/auth/otp/send', { email } as SendOtpRequest);
      return response.data;
    } catch (error) {
      return this.handleApiError(error, 'Failed to send OTP.');
    }
  }

  /**
   * Verifies the OTP and completes the login/registration process.
   * On success, it saves the session tokens.
   * @param email The user's email address.
   * @param otp The one-time password received by the user.
   * @returns A service response with auth details on success.
   */
  public async verifyOtp(email: string, otp: string): Promise<ServiceResponse<AuthSuccessResponse>> {
    if (!email || !otp) {
      return { success: false, message: 'Email and OTP are required for verification.' };
    }
    try {
      const response = await this.api.post<ApiResponse<AuthSuccessResponse>>('/auth/otp/verify', { email, otp } as VerifyOtpRequest);
      
      // Save the new tokens for session management
      await this.saveTokens({
        accessToken: response.data.data.accessToken,
        refreshToken: response.data.data.refreshToken,
      });

      return response.data;
    } catch (error) {
      return this.handleApiError(error, 'OTP verification failed.');
    }
  }

  /**
   * Logs the user out by invalidating the session on the server and clearing local storage.
   * @returns A service response indicating success or failure.
   */
  public async logout(): Promise<ServiceResponse<{ message: string }>> {
    try {
      // Attempt to invalidate session on the backend
      await this.api.post('/auth/logout');
      // Clear local storage regardless of backend success for a clean client state
      await this.clearSession();
      return { success: true, data: { message: 'Logged out successfully.' } };
    } catch (error) {
      // Even if the backend call fails (e.g., token already expired), we clear local storage
      await this.clearSession();
      // We can still return a success for the client-side action
      return { success: true, data: { message: 'Logged out successfully (server response ignored).' } };
    }
  }

  /**
   * Attempts to refresh the access token using the stored refresh token.
   * This is typically called by the interceptor on a 401 error.
   * @returns A promise that resolves to the new tokens or null on failure.
   */
  private async refreshSession(): Promise<AuthTokens | null> {
    const tokens = await this.getTokens();
    if (!tokens?.refreshToken) {
      return null;
    }

    try {
      const response = await this.api.post<ApiResponse<AuthSuccessResponse>>('/auth/token/refresh', {
        refreshToken: tokens.refreshToken,
      });

      const newTokens: AuthTokens = {
        accessToken: response.data.data.accessToken,
        refreshToken: response.data.data.refreshToken,
      };

      await this.saveTokens(newTokens);
      return newTokens;
    } catch (error) {
      // Refresh failed, clear session and return null to trigger logout flow
      await this.clearSession();
      return null;
    }
  }

  // --- Wallet Management Method ---

  /**
   * Creates a new wallet for the authenticated user.
   * Requires a valid access token to be present in the session.
   * @returns A service response with wallet details on success.
   */
  public async createWallet(): Promise<ServiceResponse<WalletCreationResponse>> {
    if (!(await this.isLoggedIn())) {
      return { success: false, message: 'User not authenticated. Please log in first.' };
    }
    try {
      const response = await this.api.post<ApiResponse<WalletCreationResponse>>('/wallet/create', {});
      return response.data;
    } catch (error) {
      return this.handleApiError(error, 'Failed to create wallet.');
    }
  }

  // --- Error Handling Utility ---

  /**
   * Standardized error handler for Axios errors.
   * @param error The error object from the Axios call.
   * @param defaultMessage A fallback message if the error structure is unexpected.
   * @returns A standardized ApiErrorResponse object.
   */
  private handleApiError(error: unknown, defaultMessage: string): ApiErrorResponse {
    if (axios.isAxiosError(error) && error.response) {
      const status = error.response.status;
      const data = error.response.data as any;

      // Check for a standardized error response from the backend
      if (data && data.message) {
        return {
          success: false,
          message: data.message,
          code: data.code,
          details: data.details,
        };
      }

      // Handle common HTTP error statuses
      switch (status) {
        case 400:
          return { success: false, message: data.message || 'Bad Request: Invalid input data.' };
        case 401:
          return { success: false, message: 'Unauthorized: Invalid or expired token.' };
        case 403:
          return { success: false, message: 'Forbidden: You do not have permission to perform this action.' };
        case 404:
          return { success: false, message: 'Not Found: The requested resource does not exist.' };
        case 500:
          return { success: false, message: 'Server Error: An unexpected error occurred on the server.' };
        default:
          return { success: false, message: `API Error (${status}): ${data.message || defaultMessage}` };
      }
    } else if (axios.isAxiosError(error) && error.request) {
      // The request was made but no response was received (e.g., network error, timeout)
      return { success: false, message: 'Network Error: Could not connect to the server.' };
    } else if (error instanceof Error) {
      // A non-Axios error (e.g., in token storage)
      return { success: false, message: `Local Error: ${error.message}` };
    } else {
      // Unknown error
      return { success: false, message: defaultMessage };
    }
  }
}

// --- Example Usage (for demonstration, not part of the service class) ---
/*
// To use this service:
// 1. Install dependencies:
//    pnpm add axios @react-native-async-storage/async-storage
// 2. Import and instantiate:
//    const cdpAuthService = new CdpAuthService();

// Example flow:
async function handleLogin() {
  // 1. Send OTP
  const sendResult = await cdpAuthService.sendOtp('agent@54link.io');
  if (!sendResult.success) {
    console.error('Send OTP failed:', sendResult.message);
    return;
  }
  console.log('OTP sent successfully.');

  // 2. Verify OTP (assuming user enters '123456')
  const verifyResult = await cdpAuthService.verifyOtp('agent@54link.io', '123456');
  if (!verifyResult.success) {
    console.error('Verify OTP failed:', verifyResult.message);
    return;
  }
  console.log('Login successful. User ID:', verifyResult.data.userId);

  // 3. Check if wallet needs creation
  if (!verifyResult.data.walletCreated) {
    console.log('Wallet not found, creating...');
    const walletResult = await cdpAuthService.createWallet();
    if (walletResult.success) {
      console.log('Wallet created:', walletResult.data.walletId);
    } else {
      console.error('Wallet creation failed:', walletResult.message);
    }
  }

  // 4. Logout
  // await cdpAuthService.logout();
}
*/

File: ios-native/RemittanceApp/Services/CDPAuthService.swift
//
// CDPAuthService.swift
// 54Link Agency Banking
//
// Created by Manus AI on 2025-11-05.
// Copyright © 2025 54Link Agency Banking. All rights reserved.
//

import Foundation

// MARK: - API Endpoints and Configuration

/// A structure to hold configuration details for the CDP API.
struct CDPAPIConfig {
    static let baseURL = "https://api.nigerianremittance.com/cdp/v1"
    
    // Authentication Endpoints
    static let requestOTPEndpoint = baseURL + "/auth/otp/request"
    static let verifyOTPEndpoint = baseURL + "/auth/otp/verify"
    
    // Wallet and Session Endpoints
    static let createWalletEndpoint = baseURL + "/wallet/create"
    static let refreshSessionEndpoint = baseURL + "/session/refresh"
    static let logoutEndpoint = baseURL + "/session/logout"
}

// MARK: - Data Models

/// Represents the response from a successful OTP request.
struct OTPRequestResponse: Decodable {
    let success: Bool
    let message: String
    let otpReference: String // A reference ID to be used for OTP verification
}

/// Represents the response from a successful OTP verification and login.
struct LoginResponse: Decodable {
    let success: Bool
    let message: String
    let accessToken: String
    let refreshToken: String
    let userDidCreateWallet: Bool // Indicates if a wallet was created during this process
}

/// Represents the response from a successful wallet creation.
struct WalletCreationResponse: Decodable {
    let success: Bool
    let message: String
    let walletId: String
}

/// Represents a standardized error structure for API calls.
struct APIError: Error, Decodable {
    let code: Int
    let message: String
    
    var localizedDescription: String {
        return "Error \(code): \(message)"
    }
}

// MARK: - Service Protocol

/// Protocol defining the contract for the CDP Authentication Service.
protocol CDPAuthServiceProtocol {
    /// Requests an OTP to be sent to the provided email address.
    /// - Parameters:
    ///   - email: The user's email address.
    /// - Returns: The OTP reference string needed for verification.
    func requestOTP(email: String) async throws -> String
    
    /// Verifies the OTP and completes the login process.
    /// - Parameters:
    ///   - otpReference: The reference ID received from `requestOTP`.
    ///   - otp: The 6-digit OTP provided by the user.
    /// - Returns: The login response containing tokens and wallet status.
    func verifyOTP(otpReference: String, otp: String) async throws -> LoginResponse
    
    /// Creates a new wallet for the authenticated user.
    /// - Returns: The ID of the newly created wallet.
    func createWallet() async throws -> String
    
    /// Refreshes the user's session using the refresh token.
    /// - Returns: The new access token.
    func refreshSession() async throws -> String
    
    /// Logs out the user and invalidates the session.
    func logout() async throws
    
    /// Checks if the user is currently authenticated.
    var isAuthenticated: Bool { get }
}

// MARK: - Main Service Implementation

/// A production-ready service class for handling all CDP authentication and session management.
final class CDPAuthService: CDPAuthServiceProtocol {
    
    // MARK: - Properties
    
    /// A simple in-memory store for session tokens. In a real app, this would use Keychain.
    private var accessToken: String?
    private var refreshToken: String?
    
    /// The shared URLSession for making network requests.
    private let urlSession: URLSession
    
    /// Initializes the service with a custom URLSession (for testing) or the default shared session.
    init(urlSession: URLSession = .shared) {
        self.urlSession = urlSession
    }
    
    /// Checks if the user is currently authenticated by checking for an access token.
    var isAuthenticated: Bool {
        return accessToken != nil
    }
    
    // MARK: - Helper Methods
    
    /// Performs a generic network request and decodes the response.
    /// - Parameters:
    ///   - url: The URL for the request.
    ///   - method: The HTTP method (e.g., "POST").
    ///   - body: Optional data to be sent in the request body.
    ///   - token: Optional access token for authorization.
    /// - Returns: The decoded response object.
    private func performRequest<T: Decodable>(url: URL, method: String, body: [String: Any]? = nil, token: String? = nil) async throws -> T {
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        // Add Authorization header if a token is provided
        if let token = token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        // Add request body
        if let body = body {
            request.httpBody = try JSONSerialization.data(withJSONObject: body, options: [])
        }
        
        let (data, response) = try await urlSession.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        
        // Handle successful response (200-299)
        if (200...299).contains(httpResponse.statusCode) {
            do {
                let decoder = JSONDecoder()
                // Use keyDecodingStrategy to convert snake_case from API to camelCase in Swift
                decoder.keyDecodingStrategy = .convertFromSnakeCase
                return try decoder.decode(T.self, from: data)
            } catch {
                // Handle decoding errors
                print("Decoding Error: \(error)")
                throw error
            }
        } else {
            // Handle API errors (4xx, 5xx)
            do {
                let apiError = try JSONDecoder().decode(APIError.self, from: data)
                throw apiError
            } catch {
                // If API error decoding fails, throw a generic error
                throw URLError(.init(rawValue: httpResponse.statusCode), userInfo: [NSLocalizedDescriptionKey: "Server responded with status code \(httpResponse.statusCode)"])
            }
        }
    }
    
    // MARK: - Authentication Flow
    
    /// Requests an OTP for the given email.
    /// - Throws: `APIError` or `URLError` if the request fails.
    func requestOTP(email: String) async throws -> String {
        // 1. Input Validation
        guard isValidEmail(email) else {
            throw NSError(domain: "CDPAuthService", code: 1001, userInfo: [NSLocalizedDescriptionKey: "Invalid email format."])
        }
        
        guard let url = URL(string: CDPAPIConfig.requestOTPEndpoint) else {
            throw URLError(.badURL)
        }
        
        let body: [String: Any] = ["email": email]
        
        // 2. Perform API Request
        let response: OTPRequestResponse = try await performRequest(url: url, method: "POST", body: body)
        
        // 3. Return OTP Reference
        return response.otpReference
    }
    
    /// Verifies the OTP and logs the user in.
    /// - Throws: `APIError` or `URLError` if the request fails.
    func verifyOTP(otpReference: String, otp: String) async throws -> LoginResponse {
        // 1. Input Validation
        guard !otpReference.isEmpty, otp.count == 6, CharacterSet.decimalDigits.isSuperset(of: CharacterSet(charactersIn: otp)) else {
            throw NSError(domain: "CDPAuthService", code: 1002, userInfo: [NSLocalizedDescriptionKey: "Invalid OTP or reference provided."])
        }
        
        guard let url = URL(string: CDPAPIConfig.verifyOTPEndpoint) else {
            throw URLError(.badURL)
        }
        
        let body: [String: Any] = [
            "otp_reference": otpReference,
            "otp": otp
        ]
        
        // 2. Perform API Request
        let response: LoginResponse = try await performRequest(url: url, method: "POST", body: body)
        
        // 3. Session Management: Store tokens securely (in a real app, use Keychain)
        self.accessToken = response.accessToken
        self.refreshToken = response.refreshToken
        
        return response
    }
    
    // MARK: - Wallet Creation
    
    /// Creates a new wallet for the currently authenticated user.
    /// - Throws: `APIError`, `URLError`, or a custom error if not authenticated.
    func createWallet() async throws -> String {
        // 1. Authentication Check
        guard let token = accessToken else {
            throw NSError(domain: "CDPAuthService", code: 1003, userInfo: [NSLocalizedDescriptionKey: "User not authenticated. Please log in first."])
        }
        
        guard let url = URL(string: CDPAPIConfig.createWalletEndpoint) else {
            throw URLError(.badURL)
        }
        
        // 2. Perform API Request with Authorization
        let response: WalletCreationResponse = try await performRequest(url: url, method: "POST", token: token)
        
        // 3. Return Wallet ID
        return response.walletId
    }
    
    // MARK: - Session Management
    
    /// Refreshes the access token using the stored refresh token.
    /// - Throws: `APIError`, `URLError`, or a custom error if no refresh token is available.
    func refreshSession() async throws -> String {
        // 1. Token Check
        guard let currentRefreshToken = refreshToken else {
            throw NSError(domain: "CDPAuthService", code: 1004, userInfo: [NSLocalizedDescriptionKey: "No refresh token available. User needs to re-authenticate."])
        }
        
        guard let url = URL(string: CDPAPIConfig.refreshSessionEndpoint) else {
            throw URLError(.badURL)
        }
        
        let body: [String: Any] = ["refresh_token": currentRefreshToken]
        
        // 2. Perform API Request
        let response: LoginResponse = try await performRequest(url: url, method: "POST", body: body)
        
        // 3. Update Session Tokens
        self.accessToken = response.accessToken
        self.refreshToken = response.refreshToken // Refresh token might also be rotated
        
        return response.accessToken
    }
    
    /// Clears all session tokens and effectively logs the user out.
    func logout() async throws {
        // In a real application, you would send a request to the backend to invalidate the refresh token.
        // For this implementation, we will simulate the backend call and then clear local state.
        
        guard let token = accessToken else {
            // Already logged out or never logged in, no action needed.
            return
        }
        
        guard let url = URL(string: CDPAPIConfig.logoutEndpoint) else {
            // Even if URL is bad, we must clear local state.
            clearLocalSession()
            throw URLError(.badURL)
        }
        
        do {
            // 1. Perform API Request to invalidate session
            // The response type is irrelevant for logout, we just check for success status code.
            let _: OTPRequestResponse = try await performRequest(url: url, method: "POST", token: token)
        } catch {
            // Log the error but proceed to clear local state, as the user expects to be logged out locally.
            print("Warning: Backend logout failed: \(error.localizedDescription). Clearing local session anyway.")
        }
        
        // 2. Clear Local Session
        clearLocalSession()
    }
    
    /// Internal method to clear the locally stored tokens.
    private func clearLocalSession() {
        self.accessToken = nil
        self.refreshToken = nil
        // In a real app, this would also clear Keychain entries.
    }
    
    // MARK: - Validation Utility
    
    /// Simple email validation using a regular expression.
    /// - Parameter email: The email string to validate.
    /// - Returns: `true` if the email is valid, `false` otherwise.
    private func isValidEmail(_ email: String) -> Bool {
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format:"SELF MATCHES %@", emailRegex)
        return emailPredicate.evaluate(with: email)
    }
}

// MARK: - Example Usage (Optional)

/*
// Example of how to use the service in a ViewModel or Controller:

class AuthViewModel: ObservableObject {
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var isAuthenticated = false
    
    private let authService = CDPAuthService()
    
    func handleLogin(email: String, otp: String) async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil
        
        do {
            // 1. Request OTP
            let otpReference = try await authService.requestOTP(email: email)
            print("OTP requested successfully. Reference: \(otpReference)")
            
            // (User enters OTP)
            
            // 2. Verify OTP and Log In
            let loginResponse = try await authService.verifyOTP(otpReference: otpReference, otp: otp)
            print("Login successful. Access Token: \(loginResponse.accessToken)")
            
            isAuthenticated = authService.isAuthenticated
            
            // 3. Conditional Wallet Creation
            if !loginResponse.userDidCreateWallet {
                let walletId = try await authService.createWallet()
                print("Wallet created successfully. ID: \(walletId)")
            }
            
        } catch let apiError as APIError {
            errorMessage = "API Error: \(apiError.message)"
        } catch {
            errorMessage = "An unexpected error occurred: \(error.localizedDescription)"
        }
        
        isLoading = false
    }
    
    func handleLogout() async {
        do {
            try await authService.logout()
            isAuthenticated = authService.isAuthenticated
            print("User logged out.")
        } catch {
            errorMessage = "Logout failed: \(error.localizedDescription)"
        }
    }
    
    func refreshSessionIfNeeded() async {
        guard !isLoading else { return }
        isLoading = true
        
        do {
            let newAccessToken = try await authService.refreshSession()
            print("Session refreshed. New Access Token: \(newAccessToken)")
        } catch {
            // Handle refresh failure, e.g., force user to log in again
            print("Session refresh failed: \(error.localizedDescription)")
            isAuthenticated = false
        }
        
        isLoading = false
    }
}
*/